spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkScan.java

```diff
@@ -76,4 +109,706 @@ public void testEstimatedRowCount() throws NoSuchTableException {
+    Batch scan = builder.build().toBatch();
+
+    Assertions.assertThat(scan.planInputPartitions().length).isEqualTo(8);
+
+    // NOT GTEQ
+    builder = scanBuilder();
+
+    predicate = new Not(predicate);
+    pushFilters(builder, predicate);
+    scan = builder.build().toBatch();
+
+    Assertions.assertThat(scan.planInputPartitions().length).isEqualTo(2);
+  }
+
+  @Test
+  public void testUnpartitionedBucketLong() throws Exception {
+    createUnpartitionedTable(spark, tableName);
+
+    SparkScanBuilder builder = scanBuilder();
+
+    BucketFunction.BucketLong function = new BucketFunction.BucketLong(DataTypes.LongType);
+    UserDefinedScalarFunc udf =
+        new UserDefinedScalarFunc(
+            function.name(),
+            function.canonicalName(),
+            expressions(LiteralValue.apply(5, DataTypes.IntegerType), FieldReference.apply("id")));
+    Predicate predicate =
+        new Predicate(">=", expressions(udf, LiteralValue.apply(2, DataTypes.IntegerType)));
+    pushFilters(builder, predicate);
+    Batch scan = builder.build().toBatch();
+
+    Assertions.assertThat(scan.planInputPartitions().length).isEqualTo(10);
+
+    // NOT GTEQ
+    builder = scanBuilder();
+
+    predicate = new Not(predicate);
+    pushFilters(builder, predicate);
+    scan = builder.build().toBatch();
+
+    Assertions.assertThat(scan.planInputPartitions().length).isEqualTo(10);
+  }
+
+  @Test
+  public void testPartitionedBucketLong() throws Exception {
+    createPartitionedTable(spark, tableName, "bucket(5, id)");
+
+    SparkScanBuilder builder = scanBuilder();
+
+    BucketFunction.BucketLong function = new BucketFunction.BucketLong(DataTypes.LongType);
+    UserDefinedScalarFunc udf =
+        new UserDefinedScalarFunc(
+            function.name(),
+            function.canonicalName(),
+            expressions(LiteralValue.apply(5, DataTypes.IntegerType), FieldReference.apply("id")));
+    Predicate predicate =
+        new Predicate(">=", expressions(udf, LiteralValue.apply(2, DataTypes.IntegerType)));
+    pushFilters(builder, predicate);
+    Batch scan = builder.build().toBatch();
+
+    Assertions.assertThat(scan.planInputPartitions().length).isEqualTo(6);
+
+    // NOT GTEQ
+    builder = scanBuilder();
+
+    predicate = new Not(predicate);
+    pushFilters(builder, predicate);
+    scan = builder.build().toBatch();
+
+    Assertions.assertThat(scan.planInputPartitions().length).isEqualTo(4);
+  }
+
+  @Test
+  public void testUnpartitionedBucketString() throws Exception {
+    createUnpartitionedTable(spark, tableName);
+
+    SparkScanBuilder builder = scanBuilder();
+
+    BucketFunction.BucketString function = new BucketFunction.BucketString();
+    UserDefinedScalarFunc udf =
+        new UserDefinedScalarFunc(
+            function.name(),
+            function.canonicalName(),
+            expressions(
+                LiteralValue.apply(5, DataTypes.IntegerType), FieldReference.apply("data")));
```

Comment:
Some helpers would probably make this easier to read and construct, like `int_lit`, `field_ref`, or `to_udf`:

```java
  UserDefinedScalarFunc udf = to_udf(function, expressions(int_lit(5), ref("data")));
```

Reply:
Updated